<a href="https://colab.research.google.com/github/joaossoliveira/House-price/blob/main/House_Price1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import os
os.environ['KAGGLE_USERNAME'] = "jpso0907"
os.environ['KAGGLE_KEY'] = "3ec703ca5503f4f9e20de4fe23482a8b"

!kaggle competitions download -c home-data-for-ml-course

  0% 0.00/386k [00:00<?, ?B/s]
100% 386k/386k [00:00<00:00, 338MB/s]


In [3]:
!unzip home-data-for-ml-course.zip

Archive:  home-data-for-ml-course.zip
  inflating: data_description.txt    
  inflating: sample_submission.csv   
  inflating: sample_submission.csv.gz  
  inflating: test.csv                
  inflating: test.csv.gz             
  inflating: train.csv               
  inflating: train.csv.gz            


In [4]:
import pandas as pd

train_data = pd.read_csv('train.csv')
test_data = pd.read_csv('test.csv')

# Visualizar os dados
train_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

## 1. Imports <a id="imports"></a>

In [14]:
# Importa a biblioteca numpy, usada para manipulação de arrays e operações numéricas.
import numpy as np

# Importa a biblioteca pyplot do matplotlib, usada para criar gráficos.
from matplotlib import pyplot as plt

# Importa a biblioteca seaborn, que é uma extensão do matplotlib para criar gráficos mais atraentes e informativos.
import seaborn as sns

# Função para dividir os dados em treino, validação e teste
from sklearn.model_selection import train_test_split


# Importa a biblioteca de warnings para controlar a exibição de mensagens de aviso.
import warnings

# Ignora todas as mensagens de aviso, evitando que apareçam no terminal durante a execução do código.
warnings.filterwarnings("ignore")

# # Exibe todo o conteúdo da célula
# pd.set_option('display.max_colwidth', None)

# # Evita que quebre a linha automaticamente
# pd.set_option('display.expand_frame_repr', False)

# # Aumenta a largura do display para acomodar mais informações
# pd.set_option('display.width', 300)

IMPORTAÇÃO DOS DADOS

In [16]:
# Lista de nomes das colunas
columns = ["LotArea",
"GrLivArea",
# "Bedroom", # Removed 'Bedroom' as it's not in the dataset.
"FullBath",
"HalfBath",
"YearBuilt",
"YearRemodAdd",
"OverallQual",
"OverallCond",
"KitchenQual",
"BsmtQual",
"Neighborhood",
"MSZoning",
"Condition1",
"SaleType",
"SaleCondition",
"YrSold",
"MoSold",
"SalePrice"
]

# Ler os dados a partir do arquivo train.csv
df = pd.read_csv('train.csv', usecols=columns)  # Note a aspa simples fechando 'train.csv'

# Exibir as 3 primeiras linhas do DataFrame para verificar a estrutura dos dados
display(df.head(3))

MSZoning  LotArea Neighborhood Condition1  OverallQual  OverallCond  \
0       RL     8450      CollgCr       Norm            7            5   
1       RL     9600      Veenker      Feedr            6            8   
2       RL    11250      CollgCr       Norm            7            5   

   YearBuilt  YearRemodAdd BsmtQual  GrLivArea  FullBath  HalfBath  \
0       2003          2003       Gd       1710         2         1   
1       1976          1976       Gd       1262         2         0   
2       2001          2002       Gd       1786         2         1   

  KitchenQual  MoSold  YrSold SaleType SaleCondition  SalePrice  
0          Gd       2    2008       WD        Normal     208500  
1          TA       5    2007       WD        Normal     181500  
2          Gd       9    2008       WD        Normal     223500

  DATA SPLIT

In [18]:
# 3. Separar features e target
X = df.drop(columns=["SalePrice"])
y = df["SalePrice"]

# 4. Divisão treino/teste
try:
    # Para regressão (sem stratify)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y,
        test_size=0.30,
        random_state=42
    )

    print("\nDivisão realizada com sucesso:")
    print(f"Tamanho do treino: {X_train.shape[0]} amostras")
    print(f"Tamanho do teste: {X_test.shape[0]} amostras")
    print(f"\nPrimeiras linhas do treino:")
    display(X_train.head(3))

except Exception as e:
    print(f"\nErro durante a divisão treino/teste: {e}")



Divisão realizada com sucesso:
Tamanho do treino: 1022 amostras
Tamanho do teste: 438 amostras

Primeiras linhas do treino:


MSZoning  LotArea Neighborhood Condition1  OverallQual  OverallCond  \
135        RL    10400       NWAmes       Norm            7            6   
1452       RM     3675      Edwards       Norm            5            5   
762        FV     8640      Somerst       Norm            7            5   

      YearBuilt  YearRemodAdd BsmtQual  GrLivArea  FullBath  HalfBath  \
135        1970          1970       TA       1682         2         0   
1452       2005          2005       Gd       1072         1         0   
762        2009          2009       Gd       1547         2         1   

     KitchenQual  MoSold  YrSold SaleType SaleCondition  
135           TA       5    2008       WD        Normal  
1452          TA       5    2006       WD        Normal  
762           Gd       6    2010      Con        Normal

In [19]:
df = pd.concat([X_train,y_train],axis= 1)
df.shape

(1022, 18)

#### 4.2.3.Análise Exploratória dos Dados - Insights Iniciais
<a id="analise-exploratoria-de-dados"></a>

In [20]:
# verifica linhas duplicadas

df.duplicated().sum()

np.int64(0)

#### Análise e Insights

 - Não foram encontradas instâncias duplicadas no dataset

#### Tipo de Variáveis
<a id="tipo-de-variaveis"></a>

In [21]:
pd.DataFrame.info(df);

<class 'pandas.core.frame.DataFrame'>
Index: 1022 entries, 135 to 1126
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MSZoning       1022 non-null   object
 1   LotArea        1022 non-null   int64 
 2   Neighborhood   1022 non-null   object
 3   Condition1     1022 non-null   object
 4   OverallQual    1022 non-null   int64 
 5   OverallCond    1022 non-null   int64 
 6   YearBuilt      1022 non-null   int64 
 7   YearRemodAdd   1022 non-null   int64 
 8   BsmtQual       996 non-null    object
 9   GrLivArea      1022 non-null   int64 
 10  FullBath       1022 non-null   int64 
 11  HalfBath       1022 non-null   int64 
 12  KitchenQual    1022 non-null   object
 13  MoSold         1022 non-null   int64 
 14  YrSold         1022 non-null   int64 
 15  SaleType       1022 non-null   object
 16  SaleCondition  1022 non-null   object
 17  SalePrice      1022 non-null   int64 
dtypes: int64(11), object(7)
memory 

In [24]:
# separando em listas: variáveis numéricas e categóricas
num_vars = df.select_dtypes(include=['number']).columns.tolist()  # Seleciona as colunas com tipo numérico
num_vars = num_vars[:-1]  # Remover a target (coluna de classe ou resultado) da lista de variáveis numéricas

cat_vars = df.select_dtypes(exclude=['number']).columns.tolist()  # Seleciona as colunas que não são numéricas (categóricas)

# Exibindo os resultados
print("Variáveis Numéricas:", num_vars)  # Imprime as variáveis numéricas
print("Variáveis Categóricas:", cat_vars)  # Imprime as variáveis categóricas
print("Qtd Variáveis Numéricas:", len(num_vars))  # Imprime a quantidade de variáveis numéricas
print("Qtd Variáveis Categóricas:", len(cat_vars))  # Imprime a quantidade de variáveis categóricas
print("Variável Resposta:", df.SalePrice.dtypes)  # Imprime o tipo da variável resposta

Variáveis Numéricas: ['LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'GrLivArea', 'FullBath', 'HalfBath', 'MoSold', 'YrSold']
Variáveis Categóricas: ['MSZoning', 'Neighborhood', 'Condition1', 'BsmtQual', 'KitchenQual', 'SaleType', 'SaleCondition']
Qtd Variáveis Numéricas: 10
Qtd Variáveis Categóricas: 7
Variável Resposta: int64


#### Análise e Insights

- A base de treino contém 18 variáveis e 1022 observações

- 10 variáveis estão como numéricas (int64)
- 7 variáveis como categóricas
- E a variável resposta (SalePrice) está como int64

#### Análise de Missing e Zeros
<a id="analise-de-missing-zeros"></a>

In [25]:
# Calcular percentual de valores ausentes e zeros para cada variável

# Percentual de valores ausentes (NaN) para cada variável
missing_percent = df.isnull().mean() * 100  # O método isnull() retorna um DataFrame booleano onde True indica valores ausentes. Mean() calcula a média (porcentagem) de valores ausentes.

# Percentual de zeros para cada variável
zeros_percent = (df == 0).mean() * 100  # O código (df == 0) cria um DataFrame booleano, onde True representa zeros. Mean() calcula a porcentagem de zeros.

# Criar um DataFrame para mostrar os resultados de valores ausentes e zeros
missing_zeros_df = pd.DataFrame({
    'features': df.columns,  # Coluna com o nome das variáveis
    'missing': missing_percent,  # Coluna com percentual de valores ausentes
    'zeros': zeros_percent  # Coluna com percentual de zeros
})

# Resetando o índice para garantir que não haja problemas com duplicação de colunas
missing_zeros_df.reset_index(drop=True, inplace=True)

# Ordenar o DataFrame pelas variáveis com maior percentual de valores ausentes
missing_zeros_df = missing_zeros_df.sort_values(by='missing', ascending=False)  # Ordena o DataFrame pela coluna 'missing', de forma decrescente.

# Exibir o DataFrame corrigido com a coluna 'features' com o nome correto
missing_zeros_df = missing_zeros_df[['features', 'missing', 'zeros']]  # Selecionando apenas as colunas desejadas

# Exibir o DataFrame final
missing_zeros_df


features   missing      zeros
8        BsmtQual  2.544031   0.000000
0        MSZoning  0.000000   0.000000
1         LotArea  0.000000   0.000000
2    Neighborhood  0.000000   0.000000
4     OverallQual  0.000000   0.000000
3      Condition1  0.000000   0.000000
5     OverallCond  0.000000   0.000000
6       YearBuilt  0.000000   0.000000
7    YearRemodAdd  0.000000   0.000000
9       GrLivArea  0.000000   0.000000
10       FullBath  0.000000   0.293542
11       HalfBath  0.000000  63.013699
12    KitchenQual  0.000000   0.000000
13         MoSold  0.000000   0.000000
14         YrSold  0.000000   0.000000
15       SaleType  0.000000   0.000000
16  SaleCondition  0.000000   0.000000
17      SalePrice  0.000000   0.000000

#### Análise e Insights

- Temos variáveis com missing e zeros
- BsmtQual
- missing: 2.544031

- FullBath
- zeros: 0.293542

- HalfBath
- zeros: 63.013699

In [ ]:
#### 4.2.6. Variável Target (Creditability)
<a id="variavel-target"></a>

O objetivo do estudo é prever o valor de venda dos imóveis. De acordo com a descrição do dataset de Ames, a variável 'SalePrice' contém essa informação, representando o preço de venda de cada propriedade. No nosso estudo, esta variável foi definida como 'target' para a modelagem preditiva, onde buscaremos relacionar as características dos imóveis (como área construída, localização e qualidade) com seu valor de mercado.


In [29]:
# criando a variável target sendo 1 para mau e 0 para bom que é o formato mais comum (1 para evento e 0 para não evento)
df['target'] = np.where(df['SalePrice'] == 2,1,0).astype(int)

In [30]:
df['target'].value_counts().

target
0    1022
Name: count, dtype: int64

# Pergunta para o Professor se isso aqui acima não esta errado pq está tudo em Zero


In [31]:
df['target'].value_counts(normalize=True)
#mds ta tudo errado?
#pq eu to classificando se eu quero prever? mds socorro
#vou parar aqui pq não quero fazer tudo errado

target
0    1.0
Name: proportion, dtype: float64

In [32]:
df.head(2)

MSZoning  LotArea Neighborhood Condition1  OverallQual  OverallCond  \
135        RL    10400       NWAmes       Norm            7            6   
1452       RM     3675      Edwards       Norm            5            5   

      YearBuilt  YearRemodAdd BsmtQual  GrLivArea  FullBath  HalfBath  \
135        1970          1970       TA       1682         2         0   
1452       2005          2005       Gd       1072         1         0   

     KitchenQual  MoSold  YrSold SaleType SaleCondition  SalePrice  target  
135           TA       5    2008       WD        Normal     174000       0  
1452          TA       5    2006       WD        Normal     145000       0